pkg/pillar/cmd/downloader/download.go

```diff
@@ -0,0 +1,87 @@
+package downloader
+
+import (
+	"errors"
+	"fmt"
+	"net"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/lf-edge/eve/pkg/pillar/zedUpload"
+	"github.com/lf-edge/eve/pkg/pillar/zedcloud"
+)
+
+// perform the actual download
+func download(ctx *downloaderContext, trType zedUpload.SyncTransportType,
+	status Status, syncOp zedUpload.SyncOpType, downloadURL string,
+	auth *zedUpload.AuthInput, dpath, region string, maxsize uint64, ifname string,
+	ipSrc net.IP, filename, locFilename string) error {
+
+	// create Endpoint
+	dEndPoint, err := ctx.dCtx.NewSyncerDest(trType, region, dpath, auth)
+	if err != nil {
+		log.Errorf("NewSyncerDest failed: %s\n", err)
+		return err
+	}
+	// check for proxies on the selected management port interface
+	proxyUrl, err := zedcloud.LookupProxy(
+		&ctx.deviceNetworkStatus, ifname, downloadURL)
+	if err == nil && proxyUrl != nil {
+		log.Infof("%s: Using proxy %s", trType, proxyUrl.String())
+		dEndPoint.WithSrcIpAndProxySelection(ipSrc, proxyUrl)
+	} else {
+		dEndPoint.WithSrcIpSelection(ipSrc)
+	}
+
+	var respChan = make(chan *zedUpload.DronaRequest)
+
+	log.Infof("%s syncOp for <%s>, <%s>, <%s>\n", trType, dpath, region, filename)
+	// create Request
+	// Round up from bytes to Mbytes
+	maxMB := (maxsize + 1024*1024 - 1) / (1024 * 1024)
+	req := dEndPoint.NewRequest(syncOp, filename, locFilename,
+		int64(maxMB), true, respChan)
+	if req == nil {
+		return errors.New("NewRequest failed")
+	}
+
+	req.Post()
+	for {
+		select {
+		case resp, ok := <-respChan:
+			if resp.IsDnUpdate() {
+				asize := resp.GetAsize()
+				osize := resp.GetOsize()
+				log.Infof("Update progress for %v: %v/%v",
+					resp.GetLocalName(), asize, osize)
+				progress := uint(0)
+				if osize != 0 {
+					percent := 100 * asize / osize
+					progress = uint(percent)
+				}
```

Comment:
Can this be made into a utility routine for resp (zedUpload.DronaRequest ) ? 
   progress = resp.getProgress()

Reply:
I see no reason why not. 

This is another example of a place where an interface would be helpful, rather than a `struct`, allowing us to substitute other things that may simulate it.

I will get the `Progress()` in, and then we can see about interfacing that as well.